python/lsst/ctrl/mpexec/cli/pipetask.py

```diff
@@ -0,0 +1,40 @@
+# This file is part of ctrl_mpexec.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import click
+
+from lsst.daf.butler.cli.butler import LoaderCLI
+from lsst.daf.butler.cli.opt import log_level_option
+
+
+class PipetaskCLI(LoaderCLI):
+
+    localCmdPkg = "lsst.ctrl.mpexec.cli.cmd"
+
+
+@click.command(cls=PipetaskCLI, context_settings=dict(help_option_names=["-h", "--help"]))
+@log_level_option()
```

Comment:
I'm not sure I understand. It was a deliberate decision in butler to move the log level before the subcommand. butler and pipetask should agree.  Why do we require that `--log-level` comes after? We can change our minds on this.

Reply:
pipetask initially implemented common options to appear before the command and command-specific options after the command (much like `git` and other tools) but that caused confusion and I had to make it uniform. Announcement for this was here: https://community.lsst.org/t/pipetask-command-line-interface-changes/3923, I don't remember where complaints about initial command line were, on slack or some ticket.

Reply:
We have to agree on how butler and pipetask works. @TallJimbo do you have a preference?

Reply:
My understanding is that other tools with subcommands (certainly including `git` - at least the way I always use it) allow options that are not specific to the subcommand to come either before or after, and if we can do that with Click, there's no problem.

I believe `argparse` requires such options to always go before the subcommand, which is a UI disaster in this case: it requires the user to remember whether some option is specific to a subcommand, even though it's also used by (i.e. "specific to"), say, 4 of the 5 _other_ subcommands, and hence it's really demanding that the user know the interface of all of the subcommands they _aren't_ trying to call.  Telling `argparse` that all options are specific to subcommands and hence must go after the subcommand was a workaround for this: it's at least not hard to remember "all options go after the subcommand".

Do we know if Click has the same behavior as `argparse`?  If so, I think we have no choice but to do the same workaround, even if it means fighting Click, unless we can make the case that some options are so completely universal that it's easy to remember that they always must go before the subcommand.  But I hope it doesn't come to that.

Reply:
On the first implementation of the butler command we had `--log-level` after the subcommand because that was the default click behavior. We made @n8pease change it to make it before because it would always be there. If you prefer it to always come after we can change butler.

Reply:
Is there a way to tell Click to allow an option in either place?

Reply:
> is there a way to tell Click to allow an option in either place?
Kind of: we can allow the command (`pipetask`) and subcommands (e.g. `build`) to both take an option. If those options need to be mutually exclusive then we have to hand roll that in implementation. 

But, I had thought of `--log-level` and `butler --help` as being specific to the `butler` (or `pipetask`) command. `butler --help` definitely is, and we do support `butler <subcommand> --help`. As you'd expect, the later results in totally different output from the former. `butler --help` will prevent any further subcommands from running (it prints help and exits). The logging setting will affect logging from `butler` execution as well as execution of the subcommand it calls, it didn't seem like we needed subcommands to specify their own help level.

I don't think it makes sense for some subcommand options to go before the subcommand on the CLI and some to go after it. When there are common subcommands we define each in a shareable way, and it is very simple to add a shared option to the subcommand that accepts it, making it so that all options can go after the subcommand.

Reply:
if every one else thinks it should be `butler command --log-level` then change it to that.

Reply:
> allow options that are not specific to the subcommand to come either before or after, and if we can do that with Click, there's no problem.

Click does not seem to want to support e.g. `butler create ~/foo --log-level DEBUG` in an automatic sort of way. There might be a way to code around this but if possible I think we should not fight it.

Reply:
FWIW, I'm definitely not proposing allowing subcommand-specific options before the subcommand (just the converse).  And I agree `--help` currently works the way it should.

`--log-level` is one of the few things I might imagine being fine to require before the subcommand, as it could be sufficiently global.  I'm not sure it actually is that global, and that's a question of whether logging control for task execution is really the same as task control for butler logging (and really whether it's just one logging control for everything).  If it is, fine - but if not we may need to actually split this up into e.g. `--butler-log-level LEVEL` or `--task-log-level LEVEL` or something more general like `--log-level NAME=LEVEL`, before we think about what goes before or after the subcommand.